### ISSUE TITLE
Unable to make connection to unauthed Redis if REDIS_URL contains password

### ISSUE THREAD

**Issue by dv** [redis/redis-rb#831]
_This is a more annoying version of https://github.com/redis/redis-rb/issues/679_

```ruby
ENV["REDIS_URL"] = "redis://test:test@localhost:6379"
Redis.new(password: nil).ping
=> Redis::CommandError: ERR Client sent AUTH, but no password is set
```

So far I've been unable to find a (possibly hacky) way to force the Redis client to not use a password. Even if it's possible to do, it would not be very useful to the other gems which depend on this gem, that just accept a Redis configuration hash and set up their own Redis connection.

It seems it uses a value from an environment variable without any way to override it. Here's where we ran into trouble:

- We have two Redis servers, one custom hosted Redis server protected through a password which stores our persistent data. This one is set in the `REDIS_URL`.
- And another Redis server, which is actually a Google Cloud Memory Store compatible with Redis, to store our Rails cache. This one uses a completely different configuration from `REDIS_URL`.

The Google Cloud Memory Store does not support setting a password so we cannot work around this bug by configuring it to also have a password.

The result is that the Rails cache connection tries to connect with the Google Cloud Memory Store as it's configured to do, but uses the password set in `REDIS_URL` for the other Redis database.

The faulty code is probably here: https://github.com/redis/redis-rb/blob/213ed1f3f41c96f8a2e3f63bd56c8c412323254a/lib/redis/client.rb#L437

**Comment by byroot** [comment]
As discussed in #679 (and some other issues) we can't really change much of that without severely breaking backward compatibilty.

That being said I just merged https://github.com/redis/redis-rb/commit/2a8606eacaec8b43eef6d38452a216a2fb46d548 so now passing `password: nil` will make it so that the password in `REDIS_URL` won't be inherited.

This might have consequences for some corner cases, but it's likely for the best going forward.

**Comment by byroot** [comment]
Actually that broke quite a bit of tests, and after digging the reason why it was implemented like this, I'm afraid it would break much more users that I'd like: https://github.com/redis/redis-rb/issues/255#issuecomment-7108883

So instead I merged d837a0b7b46fc1cded31c8c4849a9bb286905aca

TL;DR; by passing `url: false` the client will ignore the `REDIS_URL` env var. Hope this solves your issue.

**Comment by dv** [comment]
Thank you, that should work.

**Comment by AfolabiOlaoluwa** [comment]
Thank you very much. After upgrading to Ruby 2.2.10 I get this error below:

```
Redis::CommandError - ERR invalid expire time in setex:
  redis (3.0.7) lib/redis/client.rb:97:in `call'
  redis (3.0.7) lib/redis.rb:691:in `block in setex'
  redis (3.0.7) lib/redis.rb:37:in `block in synchronize'
  /Users/mac/.rubies/ruby-2.2.10/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
  redis (3.0.7) lib/redis.rb:37:in `synchronize'
  redis (3.0.7) lib/redis.rb:690:in `setex'
  redis-store (1.1.4) lib/redis/store/interface.rb:17:in `setex'
  redis-store (1.1.4) lib/redis/store/marshalling.rb:13:in `block in setex'
  redis-store (1.1.4) lib/redis/store/marshalling.rb:29:in `_marshal'
  redis-store (1.1.4) lib/redis/store/marshalling.rb:13:in `setex'
  redis-store (1.1.4) lib/redis/store/ttl.rb:6:in `set'
  redis-store (1.1.4) lib/redis/store/marshalling.rb:5:in `block in set'
  redis-store (1.1.4) lib/redis/store/marshalling.rb:29:in `_marshal'
  redis-store (1.1.4) lib/redis/store/marshalling.rb:5:in `set'
  redis-rack (1.5.0) lib/rack/session/redis.rb:41:in `block in set_session'
  redis-rack (1.5.0) lib/rack/session/redis.rb:55:in `with_lock'
  redis-rack (1.5.0) lib/rack/session/redis.rb:40:in `set_session'
  rack (1.5.5) lib/rack/session/abstract/id.rb:342:in `commit_session'
  rack (1.5.5) lib/rack/session/abstract/id.rb:226:in `context'
  rack (1.5.5) lib/rack/session/abstract/id.rb:220:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/cookies.rb:486:in `call'
  activerecord (4.0.13) lib/active_record/query_cache.rb:36:in `call'
  activerecord (4.0.13) lib/active_record/connection_adapters/abstract/connection_pool.rb:626:in `call'
  activerecord (4.0.13) lib/active_record/migration.rb:373:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
  activesupport (4.0.13) lib/active_support/callbacks.rb:373:in `_run__4426420521352090713__call__callbacks'
  activesupport (4.0.13) lib/active_support/callbacks.rb:80:in `run_callbacks'
  actionpack (4.0.13) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/reloader.rb:64:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/remote_ip.rb:76:in `call'
  better_errors (1.1.0) lib/better_errors/middleware.rb:84:in `protected_app_call'
  better_errors (1.1.0) lib/better_errors/middleware.rb:79:in `better_errors_call'
  better_errors (1.1.0) lib/better_errors/middleware.rb:56:in `call'
  bugsnag (6.11.0) lib/bugsnag/integrations/rack.rb:46:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  railties (4.0.13) lib/rails/rack/logger.rb:38:in `call_app'
  railties (4.0.13) lib/rails/rack/logger.rb:20:in `block in call'
  activesupport (4.0.13) lib/active_support/tagged_logging.rb:68:in `block in tagged'
  activesupport (4.0.13) lib/active_support/tagged_logging.rb:26:in `tagged'
  activesupport (4.0.13) lib/active_support/tagged_logging.rb:68:in `tagged'
  railties (4.0.13) lib/rails/rack/logger.rb:20:in `call'
  quiet_assets (1.0.2) lib/quiet_assets.rb:18:in `call_with_quiet_assets'
  request_store (1.1.0) lib/request_store/middleware.rb:8:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/request_id.rb:21:in `call'
  rack (1.5.5) lib/rack/methodoverride.rb:21:in `call'
  rack (1.5.5) lib/rack/runtime.rb:17:in `call'
  lib/middleware/security.rb:11:in `call'
  rack (1.5.5) lib/rack/lock.rb:17:in `call'
  actionpack (4.0.13) lib/action_dispatch/middleware/static.rb:84:in `call'
  lib/middleware/i18n_js.rb:9:in `call'
  rack (1.5.5) lib/rack/sendfile.rb:112:in `call'
  rack-cors (1.0.2) lib/rack/cors.rb:97:in `call'
  railties (4.0.13) lib/rails/engine.rb:511:in `call'
  railties (4.0.13) lib/rails/application.rb:97:in `call'
  railties (4.0.13) lib/rails/railtie/configurable.rb:30:in `method_missing'
  rack-reverse-proxy (0.12.0) lib/rack_reverse_proxy/roundtrip.rb:19:in `call'
  rack-reverse-proxy (0.12.0) lib/rack_reverse_proxy/middleware.rb:25:in `call'
  rack (1.5.5) lib/rack/lock.rb:17:in `call'
  rack (1.5.5) lib/rack/content_length.rb:14:in `call'
  rack (1.5.5) lib/rack/handler/webrick.rb:60:in `service'
  /Users/mac/.rubies/ruby-2.2.10/lib/ruby/2.2.0/webrick/httpserver.rb:138:in `service'
  /Users/mac/.rubies/ruby-2.2.10/lib/ruby/2.2.0/webrick/httpserver.rb:94:in `run'
  /Users/mac/.rubies/ruby-2.2.10/lib/ruby/2.2.0/webrick/server.rb:294:in `block in start_thread'
```

How do you manage to fix this?
